xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-plugins/src/main/webjar/xwiki-macro/macroEditor.js

```diff
@@ -358,73 +115,136 @@ define('macroParameterTreeDisplayer', ['jquery', 'l10n!macroEditor'], function($
+      case 'GROUP': return displayGroup(parametersMap, node);
+      case 'PARAMETER': return displayMacroParameter(node);
     }
   },
 
-  macroParameterGroupTemplate = 
-    '<li class="macro-parameter-group">' +
-      '<ul class="nav nav-tabs" role="tablist">' +
-        '<li class="active" role="presentation">' +
-          '<a class="macro-parameter-group-name" href="#groupId" aria-controls="groupId" role="tab"></a>' +
-        '</li>' +
-      '</ul>' +
-      '<div class="tab-content">' +
-        '<ul id="groupId" class="macro-parameter-group-members tab-pane active" role="tabpanel"></ul>' +
+  macroParameterGroupOptionalsTemplate =
+      '<div class="macro-parameter-group-optionals single-choice">' +
+        '<ul class="nav nav-tabs" role="tablist">' +
+          '<li class="active" role="presentation">' +
+            '<a class="macro-parameter-group-name" href="#groupId" aria-controls="groupId" role="tab"></a>' +
+          '</li>' +
+        '</ul>' +
+        '<div class="tab-content">' +
+          '<div id="groupId" class="macro-parameter-group-members tab-pane active" role="tabpanel"></div>' +
+        '</div>' +
+      '</div>',
+
+  displayOptionalNodes = function(parametersMap, optionalNodeList) {
+    let output = $(macroParameterGroupOptionalsTemplate);
+    let tabs = output.find('ul[role="tablist"]');
+    let tabTemplate = tabs.children().remove();
+    let tabPanels = output.find('.tab-content');
+    let tabPanelTemplate = tabPanels.children().remove();
+    optionalNodeList.forEach(function(nodeKey, index) {
+      let childNode = parametersMap[nodeKey];
+      let tab = tabTemplate.clone().appendTo(tabs);
+      let tabPanel = tabPanelTemplate.clone().appendTo(tabPanels);
+      // Some of the child nodes might be hidden so we will activate the first visible tab at the end.
+      tab.add(tabPanel).removeClass('active').toggleClass('hidden', !!childNode.hidden);
+      fillNodeTab(parametersMap, childNode, tab.children().first(), tabPanel);
+    });
+    // Activate the first visible tab.
+    let activeTab = tabs.children().not('.hidden').first();
+    let activeTabPanel = tabPanels.children().not('.hidden').first();
+    activeTab.add(activeTabPanel).addClass('active');
+    toggleMacroParameterGroupVisibility(output);
+    // we remove all occurrences of feature title since we already have the tabs name.
+    output.find('.feature-container .panel-heading').remove();
+    return output;
+  },
+
+  displayGroup = function(parametersMap, groupNode) {
+    return displayFeature(parametersMap, groupNode, groupNode.featureOnly);
+  },
+
+  macroFeatureContainerTemplate =
+    '<div class="feature-container panel panel-default" id="feature-{{featureName}}">' +
+      '<div class="panel-heading">' +
+        '<span class="feature-title">{{featureTitle}}</span>' +
+        '<span class="mandatory"></span>' +
       '</div>' +
-    '</li>',
+      '<div class="panel-body"></div>' +
+    '</div>',
```

Comment:
You could use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals .